src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -291,171 +299,77 @@ public BoundExpression LowerIsPattern(BoundPattern pattern, CSharpCompilation co
             }
         }
 
-        private BoundExpression MakeTypeTestAndAssignment(BoundExpression loweredTarget, BoundExpression loweredInput, TypeSymbol type)
+        private BoundExpression MakeEqual(BoundExpression loweredLiteral, BoundExpression input)
         {
-            Debug.Assert(type == loweredTarget.Type);
-
-            // If the match is impossible, we simply evaluate the input and yield false.
-            var matchConstantValue = MatchConstantValue(loweredInput, type, false);
-            if (matchConstantValue == false)
-            {
-                return _factory.MakeSequence(loweredInput, _factory.Literal(false));
-            }
-
-            // It is possible that the input value is already of the correct type, in which case the pattern
-            // is irrefutable, and we can just do the assignment and return true (or perform the null test).
-            if (matchConstantValue == true)
+            if (loweredLiteral.Type.SpecialType == SpecialType.System_Double && Double.IsNaN(loweredLiteral.ConstantValue.DoubleValue) ||
+                loweredLiteral.Type.SpecialType == SpecialType.System_Single && Single.IsNaN(loweredLiteral.ConstantValue.SingleValue))
             {
-                BoundExpression convertedInput;
-                if (loweredInput.Type.IsNullableType())
+                // NaN must be treated specially, as operator== and .Equals() disagree.
+                Debug.Assert(loweredLiteral.Type == input.Type);
+                var condition = _factory.InstanceCall(loweredLiteral, "Equals", input);
```

Comment:
>var condition = _factory.InstanceCall(loweredLiteral, "Equals", input); [](start = 16, length = 71)

I think we should use SpecialMember infrastructure to target specific helpers.

Reply:
In that case I'll use `double.IsNaN(input)` and `float.IsNaN(input)`...

Reply:
Addressed in https://github.com/dotnet/roslyn/pull/24059/commits/71a87eca6c65dbda7ec58daa6d4cfdc7cd310dbc of https://github.com/dotnet/roslyn/pull/24059